### PR TITLE
Completer py api

### LINKED ISSUE #3972: Python API for auto-completers
## xonfig

<details>

```
+------------------+----------------------+
| xonsh            | 0.9.24               |
| Git SHA          | 74543ea9             |
| Commit Date      | Oct 10 15:12:47 2020 |
| Python           | 3.8.6                |
| PLY              | 3.11                 |
| have readline    | True                 |
| prompt toolkit   | 3.0.3                |
| shell type       | prompt_toolkit       |
| pygments         | 2.7.2                |
| on posix         | True                 |
| on linux         | True                 |
| distro           | manjaro              |
| on darwin        | False                |
| on windows       | False                |
| on cygwin        | False                |
| on msys2         | False                |
| is superuser     | False                |
| default encoding | utf-8                |
| xonsh encoding   | utf-8                |
| encoding errors  | surrogateescape      |
| on jupyter       | False                |
| jupyter kernel   | None                 |
| xontrib 1        | abbrevs              |
| xontrib 2        | argcomplete          |
| xontrib 3        | autovox              |
| xontrib 4        | back2dir             |
| xontrib 5        | cmd_done             |
| xontrib 6        | hist_navigator       |
| xontrib 7        | jedi                 |
| xontrib 8        | kitty                |
| xontrib 9        | notifypy             |
| xontrib 10       | pdb                  |
| xontrib 11       | prompt_ret_code      |
| xontrib 12       | vox                  |
| xontrib 13       | voxapi               |
+------------------+----------------------+
```

</details>

## Expected Behavior

for registering or removing completers one has to fiddle with `__xonsh__.completers` dictionary. It should have equivalent API as `completer` command.

#3965 

## For community
⬇️  **Please click the 👍 reaction instead of leaving a `+1` or 👍  comment**

### PR DESCRIPTION
**Pull request by jnoortheen**
fixes #3972

### CODE REVIEW

File: news/completer-py-api.rst [hunk 1]
```diff
@@ -0,0 +1,27 @@
+**Added:**
+
+* index for history's sqlite-DB
+* support passing style from RichCompleter to PTK's Completer
+* ``xonsh.cli_utils`` to create cli from functions easily.
+* Python API for completer command with ``xonsh.completer`` module functions.
```
Comment (gforsyth) [inline news/completer-py-api.rst:6]: Are all of these related to this PR or are some leftover from other PRs?

File: xonsh/completers/_aliases.py [hunk 2]
```diff
@@ -1,190 +1,119 @@
...
+        err = "The name %s is already a registered completer function." % name
     else:
-        name = args[0]
-        func_name = args[1]
-        if name in builtins.__xonsh__.completers:
-            err = ("The name %s is already a registered " "completer function.") % name
+        if func_name in xsh.ctx:
+            func = xsh.ctx[func_name]
+            if not callable(func):
+                err = "%s is not callable" % func_name
```
Comment (gforsyth) [inline xonsh/completers/_aliases.py:70]:
```suggestion
                err = f"{func_name} is not callable"
```

File: xonsh/completers/_aliases.py [hunk 3]
```diff
@@ -1,190 +1,119 @@
...
     err = None
-    if len(args) not in {2, 3}:
-        err = (
-            "completer add takes either 2 or 3 arguments.\n"
-            "For help, run:  completer help add"
-        )
+    func_name = func
+    xsh = builtins.__xonsh__  # type: ignore
+    if name in xsh.completers:
+        err = "The name %s is already a registered completer function." % name
```
Comment (gforsyth) [inline xonsh/completers/_aliases.py:65]:
```suggestion
        err = f"The name {name} is already a registered completer function."
```

File: xonsh/completers/completer.py [hunk 4]
```diff
@@ -40,3 +50,65 @@ def complete_completer(prefix, line, start, end, ctx):
         else:
             raise StopIteration
     return {i for i in possible if i.startswith(prefix)}
+
+
+def add_one_completer(name, func, loc="end"):
+    new = collections.OrderedDict()
```
Comment (gforsyth) [inline xonsh/completers/completer.py:56]: We don't need to necessarily tackle this just now, but I wonder if we even need `OrderedDict` here anymore now that dictionary order is guaranteed?

File: xonsh/completers/completer.py [hunk 5]
```diff
@@ -40,3 +50,65 @@ def complete_completer(prefix, line, start, end, ctx):
...
+
+    Parameters
+    ----------
+    name:
+        NAME is a unique name of a completer (run "completer list" to see the current
+        completers in order)
+    """
+    err = None
+    if name not in xsh_session.completers:
+        err = "The name %s is not a registered " "completer function." % name
```
Comment (gforsyth) [inline xonsh/completers/completer.py:109]:
```suggestion
        err = f"The name {name} is not a registered completer function."
```

File: xonsh/history/sqlite.py [hunk 6]
```diff
@@ -67,6 +67,13 @@ def _xh_sqlite_create_history_table(cursor):
         except sqlite3.OperationalError:
             pass
 
+        # add index on inp. since we query when erasedups is True
+        cursor.execute(
+            f"""\
+CREATE INDEX IF NOT EXISTS  idx_inp_history
+ON {XH_SQLITE_TABLE_NAME}(inp);"""
+        )
+
```
Comment (gforsyth) [inline xonsh/history/sqlite.py:76]: Maybe a few commits that belonged to a different PR got pulled in here?

**Review by gforsyth — CHANGES_REQUESTED** [review]
Thanks for putting this in @jnoortheen !  overall looks good -- added a few suggestions and I think we should separate out the sqlite fixes into a separate PR

File: news/completer-py-api.rst (hunk 1 above)
Comment (jnoortheen) [reply news/completer-py-api.rst:6]: 1 & 2. required them when trying to create a history completer. 
3 & 4. cli-utils and completer are both related. 

If you want, I can create a separate PR. but these are contained within their respective commits.

File: xonsh/completers/completer.py (hunk 4 above)
Comment (jnoortheen) [reply xonsh/completers/completer.py:56]: the orderness was advertised in py3.6 I think. But I am not sure we can depend on it and it won't change in future. The other point, could you explain ?

File: xonsh/history/sqlite.py (hunk 6 above)
Comment (jnoortheen) [reply xonsh/history/sqlite.py:76]: Yes probably the news items 1&2 could go into a separate PR. but due to some laziness :)  hope you don't mind it. the commits are contained though with proper message.

File: xonsh/completers/completer.py (hunk 4 above)
Comment (gforsyth) [reply xonsh/completers/completer.py:56]: It was advertised in Python 3.6 as an implementation detail, but it was added to the language spec in 3.7, it isn't going anywhere

Comment (jnoortheen) [reply xonsh/completers/completer.py:56]: still it is better to have this. Also I haven't added this one, I just renamed the function name. 

https://stackoverflow.com/questions/50872498/will-ordereddict-become-redundant-in-python-3-7#:~:text=No%20it%20won't%20become,an%20order%20dependent%20method%2C%20OrderedDict.

File: news/completer-py-api.rst (hunk 1 above)
Comment (gforsyth) [reply news/completer-py-api.rst:6]: that's related enough for me.  thanks for clarifying!

**Comment by gforsyth** [comment]
thanks @jnoortheen !!

**Pull request merged**